File: breathing-animation/math/clojure/breathing.clj
(ns breathing
  (:require [clojure.string :as str]))

;;; The Recursive Breath: Clojure
;;; "Simple Made Easy"

;; 1. The Serene Palette (SWEBOK v4) - Defined as Immutable Map
(def palette
  {:emerald "\u001b[38;2;52;211;153m"
   :blue    "\u001b[38;2;96;165;250m"
   :rose    "\u001b[38;2;251;113;133m"
   :reset   "\u001b[0m"})

;; 2. The Cycle Data (Vector of Maps)
(def cycle
  [{:label "Inhale... 🌿" :duration 4.0 :color (:emerald palette)}
   {:label "Hold... ☁️"   :duration 7.0 :color (:blue    palette)}
   {:label "Exhale... 🌸" :duration 8.0 :color (:rose    palette)}])

;; 3. The Function (Pure Process)
(defn breathe-phase [phase]
  (let [{:keys [label duration color]} phase]
    (println (str color label " (" duration "s)" (:reset palette)))
    (Thread/sleep (long (* duration 1000)))))

;; 4. The Infinite Recursion (loop/recur)
;; Unlike standard recursion, 'recur' does not consume stack space.
(defn start-breathing []
  (println "Starting The Recursive Breath (Clojure JVM)...")
  (loop [phases (cycle cycle)] ;; Infinite sequence from specific cycle
    (let [current-phase (first phases)]
      (breathe-phase current-phase)
      (recur (rest phases))))) ;; Tail-call optimization

;; Execution
(start-breathing)

File: breathing-animation/math/elixir/Breathing.exs
# The Concurrent Breath: Elixir
# "Everything is a process."

defmodule Breathing do
  # 1. The Serene Palette (Pattern Matching Constants)
  def color(:emerald), do: "\u001b[38;2;52;211;153m"
  def color(:blue),    do: "\u001b[38;2;96;165;250m"
  def color(:rose),    do: "\u001b[38;2;251;113;133m"
  def color(:reset),   do: "\u001b[0m"

  # 2. The Actor (The "Loop")
  # The recursive function that receives messages and transitions state.
  def loop do
    receive do
      {:phase, label, duration, color_atom} ->
        # Effect: Print to Shell
        IO.puts "#{color(color_atom)}#{label} (#{duration}s)#{color(:reset)}"
        
        # Effect: Sleep (Simulate Duration)
        Process.sleep(trunc(duration * 1000))
        
        # Transition: Determine next phase and send message to SELF
        next_phase = case label do
          "Inhale... 🌿" -> {:phase, "Hold... ☁️",   7.0, :blue}
          "Hold... ☁️"   -> {:phase, "Exhale... 🌸", 8.0, :rose}
          "Exhale... 🌸" -> {:phase, "Inhale... 🌿", 4.0, :emerald}
        end
        
        send(self(), next_phase)
        loop()
    end
  end

  # 3. Initialization
  def start do
    IO.puts "Starting The Concurrent Breath (Elixir BEAM)..."
    pid = spawn(fn -> loop() end)
    # Bootstrap the infinite cycle
    send(pid, {:phase, "Inhale... 🌿", 4.0, :emerald})
    
    # Keep main process alive indefinitely
    Process.sleep(:infinity)
  end
end

# Execution
Breathing.start()
